x-pack/legacy/plugins/integrations_manager/common/routes.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { PLUGIN_ID } from './constants';
+
+export const API_ROOT = `/api/${PLUGIN_ID}`;
+
+export enum RouteName {
+  API_LIST = 'API_LIST',
+  API_INFO = 'API_INFO',
+  API_ZIP = 'API_ZIP',
+  API_TGZ = 'API_TGZ',
+}
+
+type RouteNameKey = keyof typeof RouteName;
+type RouteMap = Record<RouteNameKey, RouteDef>;
+
+interface RouteDef {
+  path: string;
+  generatePath: (replacement?: string) => string;
+}
+
+const Routes: RouteMap = {
+  [RouteName.API_LIST]: {
+    path: `${API_ROOT}/list`,
+    generatePath() {
+      return this.path;
+    },
+  },
+  [RouteName.API_INFO]: {
+    path: `${API_ROOT}/package/{pkgkey}`,
+    generatePath: replacePkgkey,
+  },
+  [RouteName.API_ZIP]: {
+    path: `${API_ROOT}/package/{pkgkey}.zip`,
+    generatePath: replacePkgkey,
+  },
+  [RouteName.API_TGZ]: {
+    path: `${API_ROOT}/package/{pkgkey}.tar.gz`,
+    generatePath: replacePkgkey,
+  },
+};
+
+export function getNamedPath(key: RouteNameKey, replacement?: string): string {
+  const route = getNamedRoute(key);
+  if (route.generatePath) return route.generatePath(replacement);
+  return route.path;
+}
+
+export function getNamedRoute(key: RouteNameKey): RouteDef {
+  const route = Routes[key];
+  if (!route) throw new Error(`no route with key '${key}'`);
+  return route;
+}
+
+function replacePkgkey(this: RouteDef, pkgkey?: string): string {
+  if (!pkgkey) throw new Error('missing replacement value for pkgkey');
+  return this.path.replace('{pkgkey}', pkgkey);
```

Comment:
Will there be any problems with `this` binding if these methods are ever called inside arrow functions at some other point?

Or with `const { generatePath } = Routes[RouteName.API_LIST];` ?

Reply:
> Will there be any problems with this binding if these methods are ever called inside arrow functions at some other point?

I think we’re ok on both of those but if not we can revisit and replace those functions. 

The `generatePath` function agrees to accept some value you’d like added to the path, and return a string for the path. Each route definition can do with that as they choose in-between. 

The handlers initially all had the same `replace('{pkgkey}', pkgkey);` line. I changed it to what’s here, but it would have been fine to leave it that way or to duplicate the string used for `path` above.

> Or with const { generatePath } = Routes[RouteName.API_LIST]; ?

The contract is that the `RouteDef` will have a `generatePath` function with the signature I mentioned. However, `Routes[RouteName.API_LIST]` is a public/supported way to access a route definition. 

The modules exports two `getNamed*(key: RouteNameKey)` getters so that it’s free to make changes to the way routes are stored.

Reply:
I think what I mean is this:
```ts
const { generatePath } = getNamedRoute(RouteName.API_List);

generatePath(); // returns undefined
```
I think that's a footgun I'd like to avoid right from the start.